cdap-support-bundle/src/main/java/io/cdap/cdap/task/SupportBundleK8sHealthCheckTask.java

```diff
@@ -0,0 +1,302 @@
+import io.kubernetes.client.openapi.ApiClient;
+import io.kubernetes.client.openapi.ApiException;
+import io.kubernetes.client.openapi.apis.CoreV1Api;
+import io.kubernetes.client.openapi.models.V1Node;
+import io.kubernetes.client.openapi.models.V1NodeList;
+import io.kubernetes.client.openapi.models.V1ObjectMeta;
+import io.kubernetes.client.openapi.models.V1Pod;
+import io.kubernetes.client.openapi.models.V1PodList;
+import io.kubernetes.client.openapi.models.V1Service;
+import io.kubernetes.client.openapi.models.V1ServiceList;
+import io.kubernetes.client.util.Config;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Collects support bundle kubernetes pods, nodes, events and logs from data fusion instance.
+ */
+public class SupportBundleK8sHealthCheckTask implements SupportBundleTask {
+
+  private static final Logger LOG = LoggerFactory.getLogger(SupportBundlePipelineInfoTask.class);
+  private static final Gson GSON = new GsonBuilder().create();
+  private static final String INSTANCE_LABEL = "master.environment.k8s.instance.label";
+  private static final String DEFAULT_INSTANCE_LABEL = "cdap.instance";
+
+  private final CConfiguration cConfiguration;
+  private final File basePath;
+  private final List<NamespaceId> namespaceIdList;
+  private final RemoteHealthCheckFetcher remoteHealthCheckFetcher;
+  private ApiClient apiClient;
+  private CoreV1Api coreV1Api;
+  private V1PodList podList;
+
+  @Inject
+  public SupportBundleK8sHealthCheckTask(CConfiguration cConfiguration, File basePath,
+                                         List<NamespaceId> namespaceIdList,
+                                         RemoteHealthCheckFetcher remoteHealthCheckFetcher) {
+    this.cConfiguration = cConfiguration;
+    this.basePath = basePath;
+    this.namespaceIdList = namespaceIdList;
+    this.remoteHealthCheckFetcher = remoteHealthCheckFetcher;
+    try {
+      apiClient = Config.defaultClient();
+      coreV1Api = new CoreV1Api(apiClient);
+    } catch (Exception e) {
+      // should not happen
+      throw new RuntimeException("Exception while getting coreV1Api", e);
+    }
+  }
+
+  /**
+   * Adds kubernetes pods, nodes, events and logs into files
+   */
+  @Override
+  public void collect() throws IOException, NotFoundException, ApiException {
+    // Get the instance label to setup prefix for K8s services
+    String instanceLabel = cConfiguration.get(INSTANCE_LABEL);
+    if (instanceLabel == null) {
+      instanceLabel = DEFAULT_INSTANCE_LABEL;
+    }
+
+    File supportBundleK8sHealthCheckFolder = new File(basePath, "health-check");
+    DirUtils.mkdirs(supportBundleK8sHealthCheckFolder);
+
+    for (NamespaceId namespaceId : namespaceIdList) {
```

Comment:
Might be worth it to split this task up by service, that way we can effectively parallelize health check tasks.

Reply:
i did some investigation, for supportBundleTaskFactories which we can't predict the number of namespaces we gonna have, so I don't think it's doable